### PR TITLE
Feature: Eventbridge v2: Add tagging

### PR DESCRIPTION
**Pull request by maxhoheiser**
## Motivation
Event bridge allows for adding tags to event buses and rules: https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-tagging.html
This PR introduces the ability to create event buses and rules with tags, add and remove tags to existing buses or rules and list tags for buses and rules.

## Changes
Use most widely used pattern of handling tags from multiple other services by extend store with TagDict mapping resource arn to the dictionary of tags.
Exten test suite to cover all different options of tagging.

### CODE REVIEW

File: localstack/services/events/models.py [hunk 1]
```diff
@@ -83,11 +86,16 @@ def __post_init__(self):
...
 
+TagDict = dict[Arn, Tag]
+
 
 class EventsStore(BaseStore):
     # Map of eventbus names to eventbus objects. The name MUST be unique per account and region (works with AccountRegionBundle)
     event_buses: EventBusDict = LocalAttribute(default=dict)
 
+    # Maps resource ARN to tags
+    TAGS: TagDict = CrossRegionAttribute(default=dict)
```
Comment (joe4dev) [inline localstack/services/events/models.py:97]: We have a `TaggingService` that implements tagging transparently (noticed by looking into DMS models as one of our new implementations). Is that applicable here or does different handling based on the resource type require a special case?

File: localstack/services/events/provider.py [hunk 2]
```diff
@@ -143,6 +156,21 @@ def format_event(event: PutEventsRequestEntry, region: str, account_id: str) ->
     return formatted_event
 
 
+def get_bus_or_rule(arn: Arn) -> ResourceType:
+    parsed_arn = parse_arn(arn)
+    service = parsed_arn["resource"].split("/", 1)[0]
```
Comment (joe4dev) [inline localstack/services/events/provider.py:161]: `service` could be misleading here given that `parsed_arn["service"]` exists. `resource_type` is more accurate here following the ARN definition: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html#arns-syntax

File: localstack/services/events/provider.py [hunk 3]
```diff
@@ -143,6 +156,21 @@ def format_event(event: PutEventsRequestEntry, region: str, account_id: str) ->
     return formatted_event
 
 
+def get_bus_or_rule(arn: Arn) -> ResourceType:
```
Comment (joe4dev) [inline localstack/services/events/provider.py:159]: maybe `get_resource_type()`

File: localstack/services/events/provider.py [hunk 4]
```diff
@@ -477,6 +514,61 @@ def put_partner_events(
...
+    @handler("TagResource")
+    def tag_resource(
+        self, context: RequestContext, resource_arn: Arn, tags: TagList, **kwargs
+    ) -> TagResourceResponse:
+        # each tag key must be unique
+        # https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html#tag-best-practices
+        store = self.get_store(context)
+        resource_type = get_bus_or_rule(resource_arn)
+        if not resource_type:
+            pass  # TODO handle error for tagging not rule or event_bus
```
Comment (joe4dev) [inline localstack/services/events/provider.py:546]: any particular reason why we don't handle that? It looks like an easy case

File: localstack/services/events/provider.py [hunk 5]
```diff
@@ -477,6 +514,61 @@ def put_partner_events(
...
+        self, context: RequestContext, resource_arn: Arn, tag_keys: TagKeyList, **kwargs
+    ) -> UntagResourceResponse:
+        store = self.get_store(context)
+        resource_type = get_bus_or_rule(resource_arn)
+        if not resource_type:
+            pass  # TODO handle error for tagging not rule or event_bus
+        self._check_resource_exists(resource_arn, resource_type, store)
+        try:
+            resource_tags = store.TAGS.get(resource_arn, {})
+        except KeyError:
```
Comment (joe4dev) [inline localstack/services/events/provider.py:564]: When can a dict.get() raise a KeyError? This seems untested

File: localstack/services/events/provider.py [hunk 6]
```diff
@@ -477,6 +514,61 @@ def put_partner_events(
...
+            raise ResourceNotFoundException(f"Resource not found: '{resource_arn}'")
+        for tag_key in tag_keys:
+            try:
```
Comment (joe4dev) [inline localstack/services/events/provider.py:567]: style: `resource_tags.pop(tag_key, None)` would be more pythonic and less verbose

File: localstack/services/events/provider.py [hunk 7]
```diff
@@ -596,12 +688,21 @@ def _get_limited_dict_and_next_token(
...
+        self, resource_arn_or_name: EventBusNameOrArn | RuleArn | None
     ) -> EventBusName:
         """Return the event bus name. Input can be either an event bus name or ARN."""
-        if not event_bus_name_or_arn:
+        if not resource_arn_or_name:
+            return "default"
+        if "arn:aws:events" not in resource_arn_or_name:
+            return resource_arn_or_name
+        resource_type = get_bus_or_rule(resource_arn_or_name)
+        # TODO how to deal with / in event bus name or rule name
```
Comment (joe4dev) [inline localstack/services/events/provider.py:699]: This seems like a recurring theme (see https://github.com/localstack/localstack/pull/10789#discussion_r1600181172). Can we get that fixed before we introduce this bug at too many different places?

File: localstack/services/events/provider.py [hunk 8]
```diff
@@ -596,12 +688,21 @@ def _get_limited_dict_and_next_token(
         return limited_dict, next_token
 
     def _extract_event_bus_name(
```
Comment (joe4dev) [inline localstack/services/events/provider.py:690]: This method is getting increasingly complex. Could be a candidate for a unit test to cover edge cases easily (as long as we can ensure AWS parity).

File: tests/aws/services/events/test_events_tags.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+import pytest
+
+from localstack.testing.pytest import markers
+from localstack.utils.strings import short_uid
+from tests.aws.services.events.helper_functions import is_v2_provider
+from tests.aws.services.events.test_events import TEST_EVENT_PATTERN
+
+
+@markers.aws.validated
+@pytest.mark.parametrize("event_bus_name", ["event_bus_default", "event_bus_custom"])
```
Comment (joe4dev) [inline tests/aws/services/events/test_events_tags.py:12]: nice coverage 💯

File: tests/aws/services/events/test_events_tags.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+    with pytest.raises(aws_client.events.exceptions.ResourceNotFoundException) as error:
+        aws_client.events.untag_resource(
+            ResourceARN=resource_arn,
+            TagKeys=[tag_key_1],
+        )
+    snapshot.match("untag_not_existing_resource_error", error)
+
+
+@markers.aws.validated
+@pytest.mark.parametrize("event_bus_name", ["event_bus_default", "event_bus_custom"])
```
Comment (joe4dev) [inline tests/aws/services/events/test_events_tags.py:128]: test design note to consider for the future: There's always a trade-off between test complexity and code duplication. That looks like an example to consider whether parametrization is worth the extra complexity given the conditionals cover most of the test. Sometimes duplication or a helper for asserts can more appropriate.

**Review by joe4dev — APPROVED** [review]
The tests are quite comprehensive. Nice coverage 👍 

My main question is whether it would be feasible to reuse the tagging service (maybe not because the resource type differentiation is too specific)?
I added some suggestions to simplify the code and resolve potential smaller bugs.

File: localstack/services/events/models.py (hunk 1 above)
Comment (maxhoheiser) [reply localstack/services/events/models.py]: used TaggingService in: https://github.com/localstack/localstack/pull/10840/commits/df9b76c5dadd4ba6b2496aacfc21bc1b8bc0fe85

The only issue I see is, that the pattern other services use is to store the TagginService objects in the store, which violates a bit of what we discussed a couple of weeks ago that data and functions should be strictly separated. This comes back to the issue that the AcountRegionBundle needs additional logic that handles this separation which would make a lot of providers cleaner.

File: localstack/services/events/provider.py (hunk 2 above)
Comment (maxhoheiser) [reply localstack/services/events/provider.py]: fixed in: https://github.com/localstack/localstack/pull/10840/commits/2ccecbc2d1b61126a3528c7296862365c04b0126

File: localstack/services/events/provider.py (hunk 5 above)
Comment (maxhoheiser) [reply localstack/services/events/provider.py]: fixed with using the TaggingService

File: localstack/services/events/provider.py (hunk 6 above)
Comment (maxhoheiser) [reply localstack/services/events/provider.py]: not needed anymore due to using TaggingService

File: localstack/services/events/provider.py (hunk 7 above)
Comment (maxhoheiser) [reply localstack/services/events/provider.py:685]: Not really, I do not have any approach to fixing this, and we seam to also assume in many other providers that names will not contain /

File: localstack/services/events/provider.py (hunk 3 above)
Comment (maxhoheiser) [reply localstack/services/events/provider.py]: refactored in: https://github.com/localstack/localstack/pull/10840/commits/ab40f18492de93569b5c43202f5aac1cccfd089c

File: localstack/services/events/provider.py (hunk 4 above)
Comment (maxhoheiser) [reply localstack/services/events/provider.py]: fixed in: https://github.com/localstack/localstack/pull/10840/commits/bd3f9db9848e4296f9c672be1f388f252c0eb56b

File: localstack/services/events/models.py (hunk 1 above)
Comment (joe4dev) [reply localstack/services/events/models.py]: Very neat ✨
IMHO, that's a compromise worth taking vs. re-implementing the same thing for every service. The likelihood of a major change in tagging is very low.
